softioc/autosave.py

```diff
@@ -0,0 +1,240 @@
+        enabled: boolean which enables or disables autosave, set to True by
+            default, or False if configure not called.
+    """
+    Autosave.directory = Path(directory)
+    Autosave.backup_on_load = backup
+    Autosave.save_period = save_period
+    Autosave.enabled = enabled
+    Autosave.device_name = name
+
+
+def start_autosave_thread():
+    autosaver = Autosave()
+    worker = threading.Thread(
+        target=autosaver.loop,
+    )
+    worker.daemon = True
+    worker.start()
+    atexit.register(_shutdown_autosave_thread, autosaver, worker)
+
+
+def _shutdown_autosave_thread(autosaver, worker):
+    autosaver.stop()
+    worker.join()
+
+
+def add_pv_to_autosave(pv, name, save_val, save_fields):
+    """Configures a PV for autosave
+
+    Args:
+        pv: a PV object inheriting ProcessDeviceSupportCore
+        name: the name of the PV which is used to generate the key
+            by which the PV value is saved to and loaded from a backup,
+            this is typically the signal name without the device prefix
+        save_val: a boolean that tracks whether to save the VAL field
+            in an autosave backup
+        save_fields: a list of string names of fields associated with the pv
+            to be saved to and loaded from a backup
+    """
+    if save_val:
+        Autosave._pvs[name] = _AutosavePV(pv)
+    for field in save_fields:
+        Autosave._pvs[f"{name}.{field}"] = _AutosavePV(pv, field)
+
+
+def load_autosave():
+    Autosave._load()
+
+
+class _AutosavePV:
+    def __init__(self, pv, field=None):
+        if not field or field == "VAL":
+            self.get = pv.get
+            self.set = pv.set
+        else:
+            self.get = lambda: pv.get_field(field)
+            self.set = lambda val: setattr(pv, field, val)
+
+
+class Autosave:
+    _pvs = {}
+    _last_saved_state = {}
+    _last_saved_time = datetime.now()
+    _stop_event = threading.Event()
+    save_period = DEFAULT_SAVE_PERIOD
+    device_name = None
+    directory = None
+    enabled = False
+    backup_on_load = False
+
+    def __init__(self):
+        if not self.enabled:
+            return
+        yaml.add_representer(
+            numpy.ndarray, _ndarray_representer, Dumper=yaml.Dumper
+        )
+        if not self.device_name:
+            raise RuntimeError(
+                "Device name is not known to autosave thread, "
+                "call autosave.configure() with keyword argument name"
+            )
+        if not self.directory:
+            raise RuntimeError(
+                "Autosave directory is not known, call "
+                "autosave.configure() with keyword argument "
+                "directory"
+            )
+        if not self.directory.is_dir():
+            raise FileNotFoundError(
+                f"{self.directory} is not a valid autosave directory"
+            )
+        self._last_saved_time = datetime.now()
+
+    @classmethod
+    def _backup_sav_file(cls):
```

Comment:
Why not use `__` prefix for local names?  This is the normal Python style for member names.